crates/ghost_actor/src/ghost_tracker.rs

```diff
@@ -0,0 +1,393 @@
+                }
+                GhostTrackerToInner::Handle(request_id, data) => {
+                    self.priv_process_handle(user_data, request_id, data)?;
+                }
+                GhostTrackerToInner::Periodic(start_delay_ms, cb) => {
+                    self.priv_process_periodic(start_delay_ms, cb)?;
+                }
+            }
+        }
+        Ok(false)
+    }
+
+    /// start tracking a pending bookmark request
+    fn priv_process_bookmark(
+        &mut self,
+        request_id: RequestId,
+        entry: GhostTrackerEntry<'lt, X, T>,
+    ) -> GhostResult<()> {
+        self.pending.insert(request_id, entry);
+        Ok(())
+    }
+
+    /// match up a pending handle request with any pending bookmarks
+    fn priv_process_handle(
+        &mut self,
+        user_data: &mut X,
+        request_id: RequestId,
+        data: T,
+    ) -> GhostResult<()> {
+        match self.pending.remove(&request_id) {
+            None => return Err(GhostErrorKind::RequestIdNotFound("".to_string()).into()),
+            Some(entry) => {
+                (entry.cb)(user_data, Ok(data))?;
+            }
+        }
+        Ok(())
+    }
+
+    /// queue up a periodic processing task
+    fn priv_process_periodic(
+        &mut self,
+        start_delay_ms: u64,
+        mut cb: GhostPeriodicCb<'lt, X>,
+    ) -> GhostResult<()> {
+        let weak_user_data_clone = self.weak_user_data.clone();
+        self.sys_ref.enqueue_processor(
+            start_delay_ms,
+            Box::new(move || match weak_user_data_clone.upgrade() {
+                Some(strong_user_data) => {
+                    let mut strong_user_data = strong_user_data.lock();
+                    cb(&mut *strong_user_data)
+                }
+                None => Ok(GhostProcessInstructions::default()),
+            }),
+        )
+    }
+
+    /// if there are any expired bookmarks, clean them up
+    fn priv_process_timeouts(&mut self, user_data: &mut X) -> GhostResult<()> {
+        let mut expired = Vec::new();
+
+        let now = std::time::Instant::now();
+
+        for (request_id, entry) in self.pending.iter() {
+            if now > entry.expires {
+                expired.push(request_id.clone())
+            }
+        }
+
+        for request_id in expired {
+            match self.pending.remove(&request_id) {
+                None => (),
+                Some(entry) => {
+                    (entry.cb)(user_data, Err("timeout".into()))?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct GhostTrackerBookmarkOptions {
+    pub timeout_ms: Option<u64>,
+}
+
+impl Default for GhostTrackerBookmarkOptions {
+    fn default() -> Self {
+        Self { timeout_ms: None }
+    }
+}
+
+impl GhostTrackerBookmarkOptions {
+    pub fn set_timeout_ms(mut self, timeout_ms: u64) -> Self {
+        self.timeout_ms = Some(timeout_ms);
+        self
+    }
+}
```

Comment:
Same here. Do we really need mutating Options?

Reply:
similar

```rust
return Ok(GhostTrackerBookmarkOptions::default().set_timeout_ms(0))
```